File: lab1/TwoSum.java
import java.io.*;
import java.util.*;

import lab3.Solution;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> numMap = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];

            if (numMap.containsKey(complement)) {
                return new int[] { numMap.get(complement), i };
            }

            numMap.put(nums[i], i);
        }

        return new int[] {};
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        try {
            BufferedReader reader = new BufferedReader(new FileReader("input1.txt"));
            String[] numStrs = reader.readLine().split(" ");
            int[] nums = new int[numStrs.length];

            for (int i = 0; i < numStrs.length; i++) {
                nums[i] = Integer.parseInt(numStrs[i]);
            }

            int target = Integer.parseInt(reader.readLine());
            reader.close();

            int[] result = solution.twoSum(nums, target);

            BufferedWriter writer = new BufferedWriter(new FileWriter("output1.txt"));
            if (result.length == 2) {
                writer.write(result[0] + " " + result[1]);
            } else {
                writer.write("Hariult oldsongui");
            }
            writer.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public TreeNode sortedArrayToBST(int[] nums) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'sortedArrayToBST'");
    }
}

File: lab1/Twosum.cpp
#include <vector>
#include <unordered_map>
#include <fstream>
#include <iostream>
#include <sstream>

class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        std::unordered_map<int, int> num_map; 
        
        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];  
            
            if (num_map.find(complement) != num_map.end()) {
                return {num_map[complement], i};
            }
            num_map[nums[i]] = i;
        }
        
        return {};  
    }
};

int main() {
    std::ifstream inputFile("input.txt");
    std::ofstream outputFile("output1.txt");

    if (!inputFile.is_open()) {
        std::cerr << "Neehed aldaa garsn!" << std::endl;
        return 1;
    }
    
    if (!outputFile.is_open()) {
        std::cerr << "Haahad aldaa garsan!" << std::endl;
        return 1;
    }
    
    std::vector<int> nums;
    int target;
    
    std::string line;
    
    if (std::getline(inputFile, line)) {
        std::istringstream iss(line);
        int num;
        while (iss >> num) {
            nums.push_back(num);
        }
    }

    if (std::getline(inputFile, line)) {
        std::istringstream iss(line);
        iss >> target;
    }
    
    Solution solution;
    std::vector<int> result = solution.twoSum(nums, target);
    
    if (result.empty()) {
        outputFile << "Hariult oldoogui." << std::endl;
    } else {
        outputFile << result[0] << " " << result[1] << std::endl;
    }
    
    inputFile.close();
    outputFile.close();
    
    return 0;
}
